pkg/controller/elasticsearch/filesettings/secret.go

```diff
@@ -0,0 +1,202 @@
+		err := settings.updateState(es, *policy)
+		if err != nil {
+			return SettingsSecret{}, err
+		}
+	}
+
+	// do not increment version if hash hasn't changed
+	if current != nil && !current.hasChanged(settings) {
+		version = current.version
+		settings.Metadata.Version = current.settings.Metadata.Version
+	}
+
+	// prepare the SettingsSecret
+	settingsBytes, err := json.Marshal(settings)
+	if err != nil {
+		return SettingsSecret{}, err
+	}
+	settingsSecret := SettingsSecret{
+		Secret: corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: es.Namespace,
+				Name:      esv1.FileSettingsSecretName(es.Name),
+				Labels:    eslabel.NewLabels(es),
+				Annotations: map[string]string{
+					settingsHashAnnotationName: settings.hash(),
+				},
+			},
+			Data: map[string][]byte{
+				settingsSecretKey: settingsBytes,
+			},
+		},
+		settings: settings,
+		version:  version,
+	}
+
+	return settingsSecret, nil
+}
+
+// NewSettingsSecretFromSecret returns a new Settings Secret from a given Secret.
+func NewSettingsSecretFromSecret(secret corev1.Secret) (SettingsSecret, error) {
+	var settings Settings
+	err := json.Unmarshal(secret.Data[settingsSecretKey], &settings)
+	if err != nil {
+		return SettingsSecret{}, err
+	}
+	version, err := strconv.ParseInt(settings.Metadata.Version, 10, 64)
+	if err != nil {
+		return SettingsSecret{}, err
+	}
+
+	return SettingsSecret{
+		Secret:   secret,
+		settings: settings,
+		version:  version,
+	}, nil
+}
+
+// SettingsSecret wraps a Secret used to store File based Settings and the corresponding Settings and their version stored in it.
+type SettingsSecret struct {
+	corev1.Secret
+	settings Settings
+	version  int64
+}
+
+// hasChanged compares the hash of the given settings with the hash stored in the annotation of the Settings Secret.
+func (s SettingsSecret) hasChanged(newSettings Settings) bool {
+	return s.Annotations[settingsHashAnnotationName] != newSettings.hash()
+}
+
+// GetVersion returns the Settings version.
+func (s SettingsSecret) GetVersion() int64 {
+	return s.version
+}
+
+// GetSettings returns the Settings.
+func (s SettingsSecret) GetSettings() Settings {
+	return s.settings
+}
+
+// CanBeOwnedBy return true if the Settings Secret can be owned by the given StackConfigPolicy, either because the Secret
+// belongs to no one or because it already belongs to the given policy.
+func (s SettingsSecret) CanBeOwnedBy(policy policyv1alpha1.StackConfigPolicy) (reconciler.SoftOwnerRef, bool) {
+	currentOwner, referenced := reconciler.SoftOwnerRefFromLabels(s.Labels)
+	// either there is no soft owner
+	if !referenced {
+		return reconciler.SoftOwnerRef{}, true
+	}
+	// or the owner is already the given policy
+	policyOwner := reconciler.SoftOwnerRef{
+		Namespace: policy.Namespace,
+		Name:      policy.Name,
+		Kind:      policy.Kind,
+	}
+	canBeOwned := currentOwner.Kind == policyOwner.Kind && currentOwner.Namespace == policyOwner.Namespace && currentOwner.Name == policyOwner.Name
```

Comment:
If you are comparing the attributes one by one we don't need the `policyOwner` struct instance? You could directly compare `policy.Namespace,policy.Name,policy.Kind`?

Reply:
Of course!